### PR TITLE
Add auto-detection for any file using # or ; for comments

### PR DESCRIPTION
**Pull request by Alhadis**
Sadly, most of the files I encounter that would benefit most from this package's highlighting lack any consistent file-extension or syntax. It's quite common to see build scripts with bareword token-lists that have #-delimited comments (at least with the projects I've been delving through lately).

This PR adds a particularly nifty feature to automatically enable this grammar for any unrecognised filetype which uses `#` or `;` to start comment-lines (the actual pattern it looks for is controlled by a package setting which is, by default, `/^\s*[#;]\s/m`; An additional package setting restricts matching to files which contain `n` or more comment-lines, to avoid incorrectly highlighting a file which happens to have one `; line like this`.

The auto-matching feature is extremely conservative: it only checks editors without an existing grammar, and also makes a habit of clearing overridden paths when closing affected files. This ensures that overridden paths won't clog the project's cache, or that an affected file will take precedence over a language package the user installs later.

**Note:** The specs currently aren't doing anything, because, uh, I suck at working with Jasmine and assume every test-runner works the same way Mocha does (and the `waitsForPromise/runs` stuff has always confused me). I decided to save myself any further frustration until I was 100% sure you'd accept this, since there's a chance it might be doing *too much* for this package.

(I could also use some help figuring out why the specs are timing out on the first test when the specs are run, haha)

/cc @lee-dohm

### CODE REVIEW

**Comment by lee-dohm** [comment]
This looks great and I would definitely accept it. If you want to use Mocha instead (especially since you're using JS for the code), you can use the `atomTestRunner` system like in my soft-wrap-indicator package: https://github.com/lee-dohm/soft-wrap-indicator.

See:

* https://github.com/lee-dohm/soft-wrap-indicator/blob/master/package.json#L32
* https://github.com/lee-dohm/soft-wrap-indicator/blob/master/package.json#L21-L24
* https://github.com/lee-dohm/soft-wrap-indicator/blob/master/spec/runner.js

**Comment by Alhadis** [comment]
Ah! Awesome, thanks for that. =)

**Comment by Alhadis** [comment]
Oh! I also updated the screenshot used in the readme file... you should also review the [readme's amended wording](https://github.com/Cutlery-Drawer/language-generic-config#generic-config-language) and let me know if anything should be rewritten. After coding for so many hours, it becomes impossible to tell what reads "naturally" (if you get what I mean…)

**Comment by lee-dohm** [comment]
Most definitely. I'll take a look soon :+1:

File: README.md [hunk 1]
```diff
@@ -7,12 +7,45 @@
...
+#rmcup=\E[?47l,
+# rs2 may need to be added
+smcup=\E[?47h,
+~~~
+
+With the package's default configuration, the above code will look like this:
+
+<img alt="Sample 2" width="433" src="sample-2.png" />
+
+You can customise how the package interprets "comment-lines" by tweaking the `automatchPattern` field in the package settings.
```
Comment (lee-dohm) [inline README.md]: `"comment-lines"` here should probably be `"comment-like" lines` for consistency.

File: README.md [hunk 2]
```diff
@@ -7,12 +7,45 @@
...
-![Sample configuration](https://raw.githubusercontent.com/lee-dohm/language-generic-config/master/sample.png)
+<img alt="Sample configuration" width="370" src="sample-1.png" />
 
-## Notes
+The package supports formats which use `#` or `;` to introduce comments.
 
-The package currently supports lines that begin with `#` or `;` as comments.
+## Auto-matching "probable" config files
+
+Many generic configs in-the-wild use no standard file extension, or any extension at all.
```
Comment (lee-dohm) [inline README.md]: `in-the-wild` shouldn't be hyphenated.

**Comment by Alhadis** [comment]
Okay, I need help again. :( I *swear* I never had this issue with [`atom-mocha`](https://www.npmjs.com/package/atom-mocha)... if I try loading this package in Mocha's `beforeEach` hook...

~~~js
atom.packages.activatePackage('language-generic-config'),
~~~

... it fails at the first test:

> Could not resolve `language-generic-config` to a package path

I pushed my changes so you can see (although I forgot to include the line in question...)

Eurgh, I feel like a dumbass. 😁

File: README.md (hunk 1 above)
Comment (Alhadis) [inline README.md]: Done! :)

**Comment by lee-dohm** [comment]
Taking a look :+1: Though it is still Sunday for me, so my brain is in low gear today :laughing:

**Comment by lee-dohm** [comment]
It looks like when running specs with the generic test runner now `ATOM_HOME` gets set to a temp directory:

```
Package Dir Paths: /var/folders/hg/1_c73qcs2bv3zh23pn9dsrjh0000gn/T/tmp-4470022hrEos34iE2/dev/packages,/var/folders/hg/1_c73qcs2bv3zh23pn9dsrjh0000gn/T/tmp-4470022hrEos34iE2/packages
```

And since `language-generic-config` isn't in or symlinked to one of those directories, it can't be found to be activated.

**Comment by Alhadis** [comment]
Possibly, although that feels awfully circuitous and inefficient... currently checking if there's a more direct way...

**Comment by Alhadis** [comment]
Ah... the solution turned out to be loading and activating the package directly. =)
~~~js
grammarPackage = atom.packages.loadPackage(join(__dirname, ".."))
~~~

Let me know if there's anything else you need.

**Comment by lee-dohm** [comment]
That definitely fixed the loading problem and was much more elegant than my solution :grinning:

I'm getting the following test failures though:

```
$ atom --test spec
GVA info: Successfully connected to the Intel plugin, offline Gen75 
Error Downloading Update: Could not get code signature for running application


  ․!!․․!

  3 passing (302ms)
  3 failing

  1) Auto-matching when opening an unrecognised file-format when the file has comments starting with `#` enables the `generic-config` grammar:

      AssertionError: expected 'text.plain.null-grammar' to equal 'text.generic-config'
      + expected - actual

      -text.plain.null-grammar
      +text.generic-config
      
      at expectScopeToBe (/Users/Lee/Source/lee-dohm/language-generic-config/spec/helpers.js:21:63)
      at waitToOpen.then (/Users/Lee/Source/lee-dohm/language-generic-config/spec/automatch.spec.js:63:11)

  2) Auto-matching when opening an unrecognised file-format when the file has comments starting with `;` enables the `generic-config` grammar:

      AssertionError: expected 'text.plain.null-grammar' to equal 'text.generic-config'
      + expected - actual

      -text.plain.null-grammar
      +text.generic-config
      
      at expectScopeToBe (/Users/Lee/Source/lee-dohm/language-generic-config/spec/helpers.js:21:63)
      at waitToOpen.then (/Users/Lee/Source/lee-dohm/language-generic-config/spec/automatch.spec.js:71:11)

  3) Auto-matching Package settings when the user disables the `enableAutomatch` setting "before each" hook for "clears all grammar overrides assigned by the package":

      AssertionError: expected 'text.plain.null-grammar' to equal 'text.generic-config'
      + expected - actual

      -text.plain.null-grammar
      +text.generic-config
      
      at expectScopeToBe (/Users/Lee/Source/lee-dohm/language-generic-config/spec/helpers.js:21:63)
      at waitToOpen.then (/Users/Lee/Source/lee-dohm/language-generic-config/spec/automatch.spec.js:99:11)
```

I'm going to keep looking at it as I can. My daughter is arriving soon for our weekly dinner so I might not be able to get very far before tomorrow.

**Comment by lee-dohm** [comment]
It looks like it works against Stable but not the latest `master`. I'll go ahead and merge it as-is and see if I can figure out what the difference is when testing against the development version of Atom.

**Comment by lee-dohm** [comment]
Thanks very much for this awesome feature :bow:

**Comment by Alhadis** [comment]
Oh, my pleasure! :D And I'm not sure where the failures were coming from. :/ Hopefully they get sorted out.

Thanks!

**Pull request merged**